controller/k8s/metadata_api.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return nil, fmt.Errorf("error configuring Kubernetes API client: %w", err)
+	}
+
+	k8sClient, err := k8s.NewAPIForConfig(config, "", []string{}, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := metadata.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	api, err := newClusterScopedMetadataAPI(k8sClient, client, resources...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, gauge := range api.gauges {
+		if err := prometheus.Register(gauge); err != nil {
+			log.Warnf("failed to register Prometheus gauge %s: %s", gauge.Desc().String(), err)
+		}
+	}
+	return api, nil
+}
+
+func newClusterScopedMetadataAPI(
+	k8sClient kubernetes.Interface,
+	metadataClient metadata.Interface,
+	resources ...APIResource,
+) (*MetadataAPI, error) {
+	sharedInformers := metadatainformer.NewFilteredSharedInformerFactory(
+		metadataClient,
+		resyncTime,
+		metav1.NamespaceAll,
+		nil,
+	)
+
+	api := &MetadataAPI{
+		Client:          k8sClient,
+		inf:             make(map[APIResource]informers.GenericInformer),
+		syncChecks:      make([]cache.InformerSynced, 0),
+		sharedInformers: sharedInformers,
+	}
+
+	for _, resource := range resources {
+		if err := api.addInformer(resource); err != nil {
+			return nil, err
+		}
+	}
+	return api, nil
+}
+
+// Sync waits for all informers to be synced.
+func (api *MetadataAPI) Sync(stopCh <-chan struct{}) {
+	api.sharedInformers.Start(stopCh)
+
+	waitForCacheSync(api.syncChecks)
+}
+
+func (api *MetadataAPI) getLister(res APIResource) (cache.GenericLister, error) {
+	inf, ok := api.inf[res]
+	if !ok {
+		return nil, fmt.Errorf("metadata informer (%v) not configured", res)
+	}
+
+	return inf.Lister(), nil
+}
+
+// GetCached returns the metadata for the supplied object type and name
+func (api *MetadataAPI) GetCached(res APIResource, name string) (*metav1.PartialObjectMetadata, error) {
+	ls, err := api.getLister(res)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := ls.Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	nsMeta, ok := obj.(*metav1.PartialObjectMetadata)
+	if !ok {
+		return nil, fmt.Errorf("couldn't convert obj %v to PartialObjectMetadata", obj)
+	}
+
+	return nsMeta, nil
+}
+
+func (api *MetadataAPI) getByNamespaceCached(res APIResource, ns, name string) (*metav1.PartialObjectMetadata, error) {
+	ls, err := api.getLister(res)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := ls.ByNamespace(ns).Get(name)
+	if err != nil {
+		return nil, err
+	}
+
+	nsMeta, ok := obj.(*metav1.PartialObjectMetadata)
+	if !ok {
+		return nil, fmt.Errorf("couldn't convert obj %v to PartialObjectMetadata", obj)
+	}
+
+	return nsMeta, nil
+}
+
+// GetByNamespaceFilteredCached returns a list of Kubernetes object references,
+// given a type, namespace, name and label selector
+func (api *MetadataAPI) GetByNamespaceFilteredCached(
+	restype APIResource,
+	ns string,
+	name string,
+	label labels.Selector,
+) ([]*corev1.ObjectReference, error) {
```

Comment:
why does this return `ObejectReference`s while the other functions return `PartialObjectMetadata`?  Should we be consistent?

Reply:
This was so because the objects returned by the metadata lister didn't have their `TypeMeta` properly filled, which was a problem when issuing an `Event` in `webhook.go`. I'll push a modification to take care of that missing info so we can return `PartialObjectMetadata` here as well :+1: